File: app/src/main/java/eu/napcode/gonoteit/repository/notes/NotesRepositoryImpl.java
package eu.napcode.gonoteit.repository.notes;

import android.annotation.SuppressLint;
import android.arch.lifecycle.LiveData;
import android.arch.lifecycle.MutableLiveData;

import java.util.ArrayList;
import java.util.List;

import javax.inject.Inject;

import eu.napcode.gonoteit.api.ApiEntity;
import eu.napcode.gonoteit.api.Note;
import eu.napcode.gonoteit.data.notes.NotesLocal;
import eu.napcode.gonoteit.data.notes.NotesRemote;
import eu.napcode.gonoteit.data.results.DeletedResult;
import eu.napcode.gonoteit.model.note.NoteModel;
import eu.napcode.gonoteit.data.results.NoteResult;
import eu.napcode.gonoteit.data.results.NotesResult;
import eu.napcode.gonoteit.repository.Resource;
import eu.napcode.gonoteit.utils.ErrorMessages;
import eu.napcode.gonoteit.utils.NetworkHelper;

public class NotesRepositoryImpl implements NotesRepository {

    private final ErrorMessages errorMessages;
    private final NotesLocal notesLocal;
    private NotesRemote notesRemote;
    private NetworkHelper networkHelper;

    MutableLiveData<Resource> resource = new MutableLiveData<>();

    @Inject
    public NotesRepositoryImpl(NotesRemote notesRemote, NotesLocal notesLocal,
                               NetworkHelper networkHelper, ErrorMessages errorMessages) {
        this.notesRemote = notesRemote;
        this.notesLocal = notesLocal;
        this.networkHelper = networkHelper;
        this.errorMessages = errorMessages;
    }

    @Override
    public NotesResult getNotes() {

        if (networkHelper.isNetworkAvailable()) {
            updateNotesFromRemote();
        } else {
            resource.postValue(Resource.error(new Throwable(errorMessages.getOfflineMessage())));
        }

        return new NotesResult(notesLocal.getNotes(), resource);
    }

    private void updateNotesFromRemote() {

        if (notesRemote.shouldFetchChangelog()) {
            getNotesChangelog();
        } else {
            getNotesFromRemote();
        }
    }

    @SuppressLint("CheckResult")
    private void getNotesChangelog() {
        //TODO Clean it up
        notesRemote.getChangelog()
                .doOnSubscribe(it -> resource.postValue(Resource.loading(null)))
//                .filter(dataResponse -> dataResponse.hasErrors())
//                .filter(dataResponse -> dataResponse.data().changelog() != null)
//                .singleOrError()
//                .map(dataResponse -> dataResponse.data().changelog())
//                .doOnSuccess(it -> notesLocal.saveChangelog(it))
//                .doOnSuccess(it -> notesRemote.saveTimestamp(it.timestamp()))
                .subscribe(
                        dataResponse -> {

                            if (dataResponse.hasErrors()) {
                                resource.postValue(Resource.success(null));
                                //TODO auth error already displayed
                                //    resource.postValue(Resource.error(dataResponse.errors().get(0)));
                            } else {
                                resource.postValue(Resource.success(null));
                            }

                            if (dataResponse.data().changelog() != null) {
                                notesLocal.saveChangelog(dataResponse.data().changelog());
                                notesRemote.saveTimestamp(dataResponse.data().changelog().timestamp());
                            }
                        },
                        error -> resource.postValue(Resource.error(error))
                );
    }

    @SuppressLint("CheckResult")
    private void getNotesFromRemote() {
        notesRemote.getNotes()
                .doOnSubscribe(it -> resource.postValue(Resource.loading(null)))
                .subscribe(
                        data -> {
                            resource.postValue(Resource.success(null));
                            notesLocal.saveEntities(data);
                        },
                        error -> resource.postValue(Resource.error(error))
                );
    }

    @Override
    public LiveData<Resource> createNote(NoteModel noteModel) {

        if (networkHelper.isNetworkAvailable()) {
            createNoteOnRemote(noteModel);
        } else {
            resource.postValue(Resource.error(new Throwable(errorMessages.getCreatingNoteNotImplementedOfflineMessage())));
        }

        return resource;
    }

    @SuppressLint("CheckResult")
    private void createNoteOnRemote(NoteModel noteModel) {
        notesRemote.createNote(noteModel)
                .doOnSubscribe(it -> resource.postValue(Resource.loading(null)))
                .filter(response -> response.data() != null)
                .filter(response -> response.data().createEntity() != null)
                .filter(response -> response.data().createEntity().ok())
                .singleOrError()
                .doOnSuccess(it -> getNotesChangelog())
                .map(dataResponse -> dataResponse.data().createEntity().entity())
                .map(entity -> (NoteModel) ((Note) entity.data()).parseNote(new ApiEntity(entity)))
                .doOnSuccess(notesLocal::saveEntity)
                .subscribe(
                        response -> resource.postValue(Resource.success(null)),
                        error -> resource.postValue(Resource.error(error))
                );
    }

    @Override
    public DeletedResult deleteNote(Long id) {
        MutableLiveData<Resource> resource = new MutableLiveData<>();

        if (networkHelper.isNetworkAvailable()) {
            deleteNoteOnRemote(id, resource);
        } else {
            resource.postValue(Resource.error(new Throwable(errorMessages.getDeletingNoteNotImplementedOfflineMessage())));
        }

        return new DeletedResult(id, resource);
    }

    @SuppressLint("CheckResult")
    private void deleteNoteOnRemote(Long id, MutableLiveData<Resource> resource) {
        notesRemote.deleteNote(id)
                .doOnSubscribe(it -> resource.postValue(Resource.loading(null)))
                .filter(response -> response.data() != null)
                .filter(response -> response.data().deleteEntity() != null)
                .filter(response -> response.data().deleteEntity().deleted())
                .singleOrError()
                .doOnSuccess(it -> getNotesChangelog())
                .subscribe(
                        response -> resource.postValue(Resource.success(null)),
                        error -> resource.postValue(Resource.error(error))
                );
    }

    @Override
    public NoteResult getNote(Long id) {

        if (networkHelper.isNetworkAvailable()) {
            getNoteFromRemote(id);
        } else {
            resource.postValue(Resource.error(new Throwable(errorMessages.getOfflineMessage())));
        }

        return new NoteResult(notesLocal.getNote(id), resource);
    }

    @SuppressLint("CheckResult")
    private void getNoteFromRemote(Long id) {
        notesRemote.getNote(id)
                .doOnSubscribe(it -> resource.postValue(Resource.loading(null)))
                .subscribe(
                        noteModel -> {
                            notesLocal.saveEntity(noteModel);
                            resource.postValue(Resource.success(null));
                        },
                        error -> resource.postValue(Resource.error(error))
                );
    }

    @Override
    public LiveData<Resource> updateNote(NoteModel noteModel) {

        if (networkHelper.isNetworkAvailable()) {
            updateNoteOnRemote(noteModel);
        } else {
            resource.postValue(Resource.error(errorMessages.getUpdatingNoteNotImplementedOfflineMessage()));
        }

        return resource;
    }


    @SuppressLint("CheckResult")
    private void updateNoteOnRemote(NoteModel noteModel) {
        notesRemote.updateNote(noteModel)
                .doOnSubscribe(it -> resource.postValue(Resource.loading(null)))
                .filter(response -> response.data() != null)
                .filter(response -> response.data().updateEntity() != null)
                .filter(response -> response.data().updateEntity().ok())
                .singleOrError()
                .map(dataResponse -> dataResponse.data().updateEntity().entity())
                .map(entity -> (NoteModel) ((Note) entity.data()).parseNote(new ApiEntity(entity)))
                .doOnSuccess(notesLocal::saveEntity)
                .doOnSuccess(it -> getNotesChangelog())
                .subscribe(
                        response -> resource.postValue(Resource.success(null)),
                        error -> resource.postValue(Resource.error(error))
                );
    }

    @Override
    public LiveData<Resource> updateFavorites(Long id) {
        List<Long> list = new ArrayList<>();
        list.add(id);
        if (networkHelper.isNetworkAvailable()) {
            notesRemote.updateFavorites(list).subscribe();
        } else {
            resource.postValue(Resource.error(errorMessages.getUpdatingNoteNotImplementedOfflineMessage()));
        }

        return resource;
    }
}

File: app/src/main/java/eu/napcode/gonoteit/repository/notes/NotesRepositoryImpl.kt
package eu.napcode.gonoteit.repository.notes

import android.annotation.SuppressLint
import android.arch.lifecycle.LiveData
import android.arch.lifecycle.MutableLiveData
import android.arch.lifecycle.Transformations
import eu.napcode.gonoteit.CreateNoteMutation
import eu.napcode.gonoteit.UpdateNoteMutation

import javax.inject.Inject

import eu.napcode.gonoteit.api.ApiEntity
import eu.napcode.gonoteit.api.Note
import eu.napcode.gonoteit.data.notes.NotesLocal
import eu.napcode.gonoteit.data.notes.NotesRemote
import eu.napcode.gonoteit.data.results.DeletedResult
import eu.napcode.gonoteit.model.note.NoteModel
import eu.napcode.gonoteit.data.results.NoteResult
import eu.napcode.gonoteit.data.results.NotesResult
import eu.napcode.gonoteit.repository.Resource
import eu.napcode.gonoteit.repository.user.UserRepository
import eu.napcode.gonoteit.utils.ErrorMessages
import eu.napcode.gonoteit.utils.NetworkHelper

class NotesRepositoryImpl @Inject
constructor(private val notesRemote: NotesRemote, private val notesLocal: NotesLocal,
            private val networkHelper: NetworkHelper, private val errorMessages: ErrorMessages,
            private val userRepository: UserRepository) : NotesRepository {
    internal var resource = MutableLiveData<Resource<*>>()

    override fun getNotes(): NotesResult {

        if (networkHelper.isNetworkAvailable) {
            updateNotesFromRemote()
        } else {
            resource.postValue(Resource.error<Any>(Throwable(errorMessages.offlineMessage)))
        }

        return NotesResult(notesLocal.notes, resource)
    }

    private fun updateNotesFromRemote() {

        if (notesRemote.shouldFetchChangelog()) {
            getNotesChangelog()
        } else {
            getNotesFromRemote()
        }
    }

    @SuppressLint("CheckResult")
    private fun getNotesChangelog() {
        //TODO Clean it up
        notesRemote.changelog
                .doOnSubscribe { it -> resource.postValue(Resource.loading<Any>(null)) }
                //                .filter(dataResponse -> dataResponse.hasErrors())
                //                .filter(dataResponse -> dataResponse.data().changelog() != null)
                //                .singleOrError()
                //                .map(dataResponse -> dataResponse.data().changelog())
                //                .doOnSuccess(it -> notesLocal.saveChangelog(it))
                //                .doOnSuccess(it -> notesRemote.saveTimestamp(it.timestamp()))
                .subscribe(
                        { dataResponse ->

                            if (dataResponse.hasErrors()) {
                                resource.postValue(Resource.success<Any>(null))
                                //TODO auth error already displayed
                                //    resource.postValue(Resource.error(dataResponse.errors().get(0)));
                            } else {
                                resource.postValue(Resource.success<Any>(null))
                            }

                            if (dataResponse.data()!!.changelog() != null) {
                                notesLocal.saveChangelog(dataResponse.data()!!.changelog()!!)
                                notesRemote.saveTimestamp(dataResponse.data()!!.changelog()!!.timestamp())
                            }
                        },
                        { error -> resource.postValue(Resource.error<Any>(error)) }
                )
    }

    @SuppressLint("CheckResult")
    private fun getNotesFromRemote() {
        notesRemote.notes
                .doOnSubscribe { it -> resource.postValue(Resource.loading<Any>(null)) }
                .subscribe(
                        { data ->
                            resource.postValue(Resource.success<Any>(null))
                            notesLocal.saveEntities(data)
                        },
                        { error -> resource.postValue(Resource.error<Any>(error)) }
                )
    }

    override fun createNote(noteModel: NoteModel): LiveData<Resource<*>> {

        if (networkHelper.isNetworkAvailable) {
            createNoteOnRemote(noteModel)
        } else {
            resource.postValue(Resource.error<Any>(Throwable(errorMessages.creatingNoteNotImplementedOfflineMessage)))
        }

        return resource
    }

    @SuppressLint("CheckResult")
    private fun createNoteOnRemote(noteModel: NoteModel) {
        notesRemote.createNote(noteModel)
                .doOnSubscribe { it -> resource.postValue(Resource.loading<Any>(null)) }
                .filter { response -> response.data() != null }
                .filter { response -> response.data()!!.createEntity() != null }
                .filter { response -> response.data()!!.createEntity()!!.ok()!! }
                .singleOrError()
                .doOnSuccess { it -> getNotesChangelog() }
                .map<CreateNoteMutation.Entity> { dataResponse -> dataResponse.data()!!.createEntity()!!.entity() }
                .map { entity -> (entity.data() as Note).parseNote(ApiEntity(entity)) as NoteModel }
                .doOnSuccess { notesLocal.saveEntity(it) }
                .subscribe(
                        { response -> resource.postValue(Resource.success<Any>(null)) },
                        { error -> resource.postValue(Resource.error<Any>(error)) }
                )
    }

    override fun deleteNote(id: Long?): DeletedResult {
        val resource = MutableLiveData<Resource<*>>()

        if (networkHelper.isNetworkAvailable) {
            deleteNoteOnRemote(id, resource)
        } else {
            resource.postValue(Resource.error<Any>(Throwable(errorMessages.deletingNoteNotImplementedOfflineMessage)))
        }

        return DeletedResult(id, resource)
    }

    @SuppressLint("CheckResult")
    private fun deleteNoteOnRemote(id: Long?, resource: MutableLiveData<Resource<*>>) {
        notesRemote.deleteNote(id)
                .doOnSubscribe { it -> resource.postValue(Resource.loading<Any>(null)) }
                .filter { response -> response.data() != null }
                .filter { response -> response.data()!!.deleteEntity() != null }
                .filter { response -> response.data()!!.deleteEntity()!!.deleted()!! }
                .singleOrError()
                .doOnSuccess { it -> getNotesChangelog() }
                .subscribe(
                        { response -> resource.postValue(Resource.success<Any>(null)) },
                        { error -> resource.postValue(Resource.error<Any>(error)) }
                )
    }

    override fun getNote(id: Long?): NoteResult {

        if (networkHelper.isNetworkAvailable) {
            getNoteFromRemote(id)
        } else {
            resource.postValue(Resource.error<Any>(Throwable(errorMessages.offlineMessage)))
        }

        return NoteResult(notesLocal.getNote(id), resource)
    }

    @SuppressLint("CheckResult")
    private fun getNoteFromRemote(id: Long?) {
        notesRemote.getNote(id)
                .doOnSubscribe { it -> resource.postValue(Resource.loading<Any>(null)) }
                .subscribe(
                        { noteModel ->
                            notesLocal.saveEntity(noteModel)
                            resource.postValue(Resource.success<Any>(null))
                        },
                        { error -> resource.postValue(Resource.error<Any>(error)) }
                )
    }

    override fun updateNote(noteModel: NoteModel): LiveData<Resource<*>> {

        if (networkHelper.isNetworkAvailable) {
            updateNoteOnRemote(noteModel)
        } else {
            resource.postValue(Resource.error<Any>(errorMessages.updatingNoteNotImplementedOfflineMessage))
        }

        return resource
    }


    @SuppressLint("CheckResult")
    private fun updateNoteOnRemote(noteModel: NoteModel) {
        notesRemote.updateNote(noteModel)
                .doOnSubscribe { it -> resource.postValue(Resource.loading<Any>(null)) }
                .filter { response -> response.data() != null }
                .filter { response -> response.data()!!.updateEntity() != null }
                .filter { response -> response.data()!!.updateEntity()!!.ok()!! }
                .singleOrError()
                .map<UpdateNoteMutation.Entity> { dataResponse -> dataResponse.data()!!.updateEntity()!!.entity() }
                .map { entity -> (entity.data() as Note).parseNote(ApiEntity(entity)) as NoteModel }
                .doOnSuccess { notesLocal.saveEntity(it) }
                .doOnSuccess { it -> getNotesChangelog() }
                .subscribe(
                        { response -> resource.postValue(Resource.success<Any>(null)) },
                        { error -> resource.postValue(Resource.error<Any>(error)) }
                )
    }

    override fun updateFavorites(id: Long): LiveData<Resource<*>> {
        val list = mutableListOf<Long>()
        list.addAll(notesLocal.getCurrentFavoriteIdsList())
        list.add(id)

        if (networkHelper.isNetworkAvailable) {
            notesRemote
                    .updateFavorites(list)
                    .doOnComplete { userRepository.updateUserFromRemote(null) }
                    .subscribe()
        } else {
            resource.postValue(Resource.error<Any>(errorMessages.updatingNoteNotImplementedOfflineMessage))
        }

        return resource
    }

    override fun getFavoriteNotes(): NotesResult {
        return NotesResult(notesLocal.getFavoriteNotes(), resource)
    }

    override fun isNoteFavorite(id: Long?): LiveData<Boolean> {
        return Transformations.map(notesLocal.getFavoriteIdsList()) {
            it.contains(id)
        }
    }
}
